File: FaceBook/ProductOfArrayExceptSelf.java
import java.util.Arrays;
import java.util.stream.IntStream;

class Solution {
  public int[] productExceptSelf(int[] nums) {
    int[] output = new int[nums.length];
    Arrays.fill(output, 1);

    IntStream.range(0, nums.length).reduce(1, (a, i) -> {
      output[i] *= a;
      return a * nums[i];
    });

    IntStream.range(0, nums.length).reduce(1, (a, i) -> {
      output[nums.length - 1 - i] *= a;
      return a * nums[nums.length - 1 - i];
    });

    return output;
  }
}

/**
 * The key here is to realize that since we cannot use division, we have to
 * concoct a way to mutliply a number of all numbers to its right and all
 * numbers to its left without the number itself. So what we will do is go left
 * to right and right to left with an accumulator, and populate the spots in the
 * output array with product of what's already in output times what the current
 * accumulated product is.
 * 
 */

File: FaceBook/productOfArrayExceptSelf.scala
object Solution {
  def productExceptSelf(nums: Array[Int]): Array[Int] = {
    val result = Array.fill(nums.length)(1)

    for (i <- 1 until nums.length) {
      result(i) = result(i - 1) * nums(i - 1)
    }
    (nums.length - 1 to 0 by -1).foldLeft(1) { (acc, i) =>
      {
        result(i) = result(i) * acc
        acc * nums(i)
      }
    }

    result
  }
}

/*
nums [ 1, 2, 3, 4 ]
        i

[1, 1, 2, 6]
answer [ 24, 12, 8, 6 ]
         i


rightProduct = 24
 */
